### PR TITLE
Implement Symfony Form CategoryType selector.

### PR DESCRIPTION
**Pull request by craigh**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| New feature?      | yes
| BC breaks?        | no
| Deprecations?     | yes
| Tests pass?       | yes
| Fixed tickets     | #2455
| Refs tickets      | -
| License           | MIT
| Doc PR            | included
| Changelog updated | yes

 - Add (move) `Zikula\CategoriesModule\Entity\AbstractCategoryAssignment` and related documentation.
 - Replaces `Zikula\Core\Doctrine\Entity\AbstractEntityCategory` (aliased for BC).
 - Add `Zikula\CategoriesModule\Form\Type\CategoriesType` for easier category usage in Symfony Forms.

### CODE REVIEW

File: src/system/CategoriesModule/Form/EventListener/CategoriesMergeCollectionListener.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    public function onBindNormData(FormEvent $event)
+    {
+        $submittedData = $event->getData();
+        $rootEntity = $event->getForm()->getParent()->getData();
+
+        $collection = new ArrayCollection();
+        foreach ($submittedData as $categoryCollectionByRegistry) {
+            foreach ($categoryCollectionByRegistry as $category) {
+                    $category->setEntity($rootEntity);
+                    $collection->add($category);
```
Comment (Guite) [inline src/system/CategoriesModule/Form/EventListener/CategoriesMergeCollectionListener.php]: Please decrease indentation here (four spaces).

File: src/system/CategoriesModule/Form/Type/CategoriesType.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        if (empty($options['entityCategoryClass']) || empty($options['module']) || empty($options['entity'])) {
+            throw new \InvalidArgumentException('empty argument!');
+        }
+
+        $registries = \CategoryRegistryUtil::getRegisteredModuleCategories($options['module'], $options['entity'], 'id');
+
+        foreach($registries as $registryId => $categoryId) {
```
Comment (Guite) [inline src/system/CategoriesModule/Form/Type/CategoriesType.php]: Missing space before opening bracket.

**Comment by Guite** [comment]
Great work!

**Comment by craigh** [comment]
> Great work!

thank you, but I cannot take all the credit. Drak did most of the original implementation. Without his work, I wouldn't have been able to find the solution. I simply had to update his code from Symfony 2.3 (or so) to more current code and then make some adjustments.

File: src/system/CategoriesModule/Form/DataTransformer/CategoriesCollectionTransformer.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        return $collection;
+    }
+    
+    public function transform($value)
+    {
+        $data = array();
+        if (empty($value)) {
+
+            return $data;
```
Comment (cmfcmf) [inline src/system/CategoriesModule/Form/DataTransformer/CategoriesCollectionTransformer.php]: There shouldn't be blank lines before return statements if the return is the only statement of the whole block.

File: src/system/CategoriesModule/Form/DataTransformer/CategoriesCollectionTransformer.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            }
+            $collection->set($regId, $subCollection);
+        }
+
+        return $collection;
+    }
+    
+    public function transform($value)
+    {
+        $data = array();
```
Comment (cmfcmf) [inline src/system/CategoriesModule/Form/DataTransformer/CategoriesCollectionTransformer.php]: You could've used [] here :P

File: src/system/CategoriesModule/Form/EventListener/CategoriesMergeCollectionListener.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+
+/**
+ * Class CategoriesMergeCollectionListener
+ * @package Zikula\CategoriesModule\Form\EventListener
+ */
+class CategoriesMergeCollectionListener implements EventSubscriberInterface
+{
+    static public function getSubscribedEvents()
```
Comment (cmfcmf) [inline src/system/CategoriesModule/Form/EventListener/CategoriesMergeCollectionListener.php]: static must be after public.

**Comment by cmfcmf** [comment]
:+1:

**Pull request merged**